File: Commons/Data/src/main/java/com/greentree/commons/data/resource/location/RootUrlResourceLocation.java
package com.greentree.commons.data.resource.location;

import java.net.MalformedURLException;
import java.net.URL;

import com.greentree.commons.data.resource.URLResource;


public final class RootUrlResourceLocation implements ResourceLocation {
	
	private static final long serialVersionUID = 1L;
	private final URL context;
	
	public RootUrlResourceLocation(URL context) {
		this.context = context;
	}
	
	@Override
	public URLResource getResource(String name) {
		URL url;
		try {
			url = new URL(context, name);
		}catch(MalformedURLException e) {
			throw new RuntimeException(e);
		}
		return new URLResource(url);
	}
	
}

File: Commons/Data/src/main/java/com/greentree/commons/data/resource/location/RootUrlResourceLocation.kt
package com.greentree.commons.data.resource.location

import com.greentree.commons.data.resource.ResourceNotFound
import com.greentree.commons.data.resource.URLFileResource
import java.net.MalformedURLException
import java.net.URI
import java.net.URL

data class RootUrlResourceLocation(
	private val context: URL,
) : ResourceLocation {

	constructor(uri: URI) : this(uri.toURL())

	constructor(uri: String) : this(URI(uri))

	override fun getResourceOrNull(name: String) = URLFileResource(
		try {
			URL(context, name)
		} catch(e: MalformedURLException) {
			throw ResourceNotFound(e)
		}
	)
}